src/com/telesign/util/TeleSignRequest.java

```diff
@@ -29,353 +29,381 @@
+     *
+     * @return A string that contains the contents of the POST Request's entity
+     *		 body.
+     */
+    public String getPostBody() {
+
+        return body;
+    }
+
+    /**
+     * Returns the set of <em>TeleSign-specific</em> Request header fields.
+     *
+     * @return A sorted key/value mapping that contains the TeleSign-specific
+     *		 Request header fields.
+     */
+    public Map<String, String> getTsHeaders() {
+
+        return ts_headers;
+    }
+
+    /**
+     * Returns the set of both the standard <u>and</u> the
+     * <em>TeleSign-specific</em> Request header fields.
+     *
+     * @return A sorted key/value mapping that contains all of the Request header fields.
+     */
+    public Map<String, String> getAllHeaders() {
+
+        TreeMap<String, String> tmp = new TreeMap<String, String>();
+
+        tmp.putAll(headers);
+        tmp.putAll(ts_headers);
+
+        return tmp;
+    }
+
+    /**
+     * Returns the set of POST Request parameters.
+     *
+     * @return A sorted key/value mapping that contains all of the POST Request parameters.
+     */
+    public Map<String, String> getAllParams() {
 
-		TreeMap<String, String> tmp = new TreeMap<String, String>();
+        return params;
+    }
 
-		tmp.putAll(headers);
-		tmp.putAll(ts_headers);
+    /**
+     * Creates and sends the REST request.
+     *
+     * @return A string containing the TeleSign web server's Response.
+     * @throws IOException A {@link java.security.SignatureException} signals that an
+     *                     error occurred while attempting to sign the Request.
+     */
+    public String executeRequest() throws IOException {
 
-		return tmp;
-	}
+        String signingString = getSigningString(customer_id);
+        String signature;
+        String url_output = "";
 
-	/**
-	 * Returns the set of POST Request parameters.
-	 *
-	 * @return A sorted key/value mapping that contains all of the POST Request parameters.
-	 */
-	public Map<String, String> getAllParams() {
+        // Create the absolute form of the resource URI, and place it in a string buffer.
+        StringBuffer full_url = new StringBuffer(base).append(resource);
 
-		return params;
-	}
+        if (params.size() > 0) {
 
-	/**
-	 * Creates and sends the REST request.
-	 *
-	 * @return A string containing the TeleSign web server's Response.
-	 * @throws IOException
-	 *			 A {@link java.security.SignatureException} signals that an
-	 *			 error occurred while attempting to sign the Request.
-	 */
-	public String executeRequest() throws IOException {
+            full_url.append("?");
+            int i = 0;
 
-		String signingString = getSigningString(customer_id);
-		String signature;
-		String url_output = "";
+            for (String key : params.keySet()) {
 
-		// Create the absolute form of the resource URI, and place it in a string buffer.
-		StringBuffer full_url = new StringBuffer(base).append(resource);
+                if (++i != 0) {
```

Comment:
If you're resorting to using indexes, just fall back to standard for loop. This code stinks and is obviously wrong. Example url it produces: 

https://rest.telesign.com/v1/phoneid/standard/15551234567?&code=001

?&  <-- ?!!